discojs/src/privacy.ts

```diff
@@ -12,13 +14,85 @@ async function frobeniusNorm(weights: WeightsContainer): Promise<number> {
   return Math.sqrt(squared[0]);
 }
 
-/** Scramble weights */
-export function addNoise(
+/** Compute the frobenius norm of weight for one layer */
+export async function layerWiseFrobeniusNorm(weight: tf.Tensor): Promise<number>{
+  const squared = await weight.square().sum().data();
+  return Math.sqrt(squared[0]);
+}
+
+/** ALDP-FL implementation */
+// Conditions need to be added for the first three epochs -> get the avg update from all of the available previous updates
+export function getClippingRadius(weightHistory: WeightHistory, defaultRadius:number|undefined = 20, windowSize: number=3, minRadius=1e-12): number | number[]{
+  // When there is no update in weightHistory yet
+  if (weightHistory.size === 0) return defaultRadius ?? 20;
+
+  const radii = Array.from(weightHistory.entries()).map(([_, norms]) => {
+    if (!norms || norms.length === 0) return defaultRadius ?? 20;
+
+    if (norms.length < windowSize){
+      // average over all available entries
+      const avg = norms.reduce((sum, n) => sum+n, 0) / norms.length;
+
+      if (avg > defaultRadius)
+        return defaultRadius
+
+      return Math.max(minRadius, avg);
+    } else{
+      const recent = norms.slice(-windowSize);
+      const avg = recent.reduce((sum, n) => sum+n, 0) / recent.length;
+
+      if (avg > defaultRadius)
+        return defaultRadius
+
+      return Math.max(minRadius, avg);
+    }
+  });
+
+  console.log('Print the clipping radius', radii);
+  return radii.length === 1 ? radii[0] : radii;
+}
+
+
+/** Add Gaussian noise per layer */
+export async function addGaussianNoise(
   weights: WeightsContainer,
-  deviation: number,
-): WeightsContainer {
-  const variance = Math.pow(deviation, 2);
-  return weights.map((w) => w.add(tf.randomNormal(w.shape, 0, variance)));
+  epsilon: number,
+  delta: number = 1e-6,
+  clippingRadius: number,  // Clipping radius of the entire weight update
+): Promise<WeightsContainer>{
+  const sens = 2*clippingRadius;
+  const sigma = (sens * Math.sqrt(2*Math.log(1.25/delta))) / epsilon; // noise calibration is the same
+  const clippedWeights = await clipNorm(weights, clippingRadius);
+  return clippedWeights.map((w) => w.add(tf.randomNormal(w.shape, 0, sigma)))
+}
+
+/** Optimized Gaussian noise using a clipping radius calculation of ALDP-FL for adaptive local differential privacy in federated learning,
+ *  https://www.nature.com/articles/s41598-025-12575-6 */
+/** Implementation of historical moving average based clipping radius calculation */
+export async function addNoise(
+  weights: WeightsContainer,
+  epsilon: number,
+  delta: number = 1e-6,
```

Comment:
optional param followed by required ones are weird to use, and as it is set everywhere, we can remove its default IMO

```suggestion
  delta: number,
```